File: java/Components/Strata.Domain.Core/src/strata/domain/core/unitofwork/AbstractUnitOfWorkState.java
// ##########################################################################
// # File Name:	AbstractUnitOfWorkState.java
// #
// # Copyright:	2017, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataPersistence Framework.
// #
// #   			The StrataPersistence Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataPersistence Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataPersistence
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata.domain.core.unitofwork;

import strata.domain.core.namedquery.INamedQuery;

import java.io.Serializable;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

/****************************************************************************
 * 
 * @author 		
 *     Sapientia Systems
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public abstract 
class AbstractUnitOfWorkState
    implements IUnitOfWorkState
{
    private final String itsName;
    
    /************************************************************************
     * Creates a new {@code AbstractUnitOfWorkState}. 
     *
     */
    protected 
    AbstractUnitOfWorkState(String name)
    {
        itsName = name;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public String
    getName()
    {
        return itsName;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<E>
    insertNew(
        AbstractUnitOfWork context,
        Class<K> keyType,
        Class<E> entityType,
        E                  newEntity)
    {
        return
            CompletableFuture.failedFuture(
                new IllegalStateException(
                    "Cannot call insert(E) from state: " + getName() + "." ));
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<E>
    updateExisting(
        AbstractUnitOfWork context,
        Class<K> keyType,
        Class<E> entityType,
        E                  existingEntity)
    {
        return
            CompletableFuture.failedFuture(
                new IllegalStateException(
                    "Cannot call update(E) from state: " +
                    getName() + "." ));
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<Void>
    removeExisting(
        AbstractUnitOfWork context,
        Class<K> keyType,
        Class<E> entityType,
        E                  existingEntity)
    {
        return
            CompletableFuture.failedFuture(
                new IllegalStateException(
                    "Cannot call remove(E) from state: " +
                    getName() + "." ));
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<Optional<E>>
    getExisting(AbstractUnitOfWork context,Class<E> type,K key)
    {
        return
            CompletableFuture.failedFuture(
                new IllegalStateException(
                    "Cannot call getUnique(Class<E>,K) from " +
                    "state: " + getName() + "." ));
    }

    /************************************************************************
     * {@inheritDoc}
     * @return
     */
    @Override
    public <E> CompletionStage<INamedQuery<E>>
    getNamedQuery(
        AbstractUnitOfWork context,
        Class<E> type,
        String queryName)
    {
        return
            CompletableFuture.failedFuture(
                new IllegalStateException(
                    "Cannot call getNamedQuery(Class<E>,String) " +
                    "from state: " + getName() + "." ));
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <K extends Serializable,E> CompletionStage<Boolean>
    hasExisting(AbstractUnitOfWork context,Class<E> type,K key)
    {
        return
            CompletableFuture.failedFuture(
                new IllegalStateException(
                    "Cannot call hasExisting(Class<E>,K) from " +
                    "state: " + getName() + "." ));
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public <E> CompletionStage<Boolean>
    hasNamedQuery(
        AbstractUnitOfWork context,
        Class<E> type,
        String queryName)
    {
        return
            CompletableFuture.failedFuture(
                new IllegalStateException(
                    "Cannot call hasNamedQuery(Class<E>,String) " +
                    "from state: " + getName() + "." ));
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    isActive(AbstractUnitOfWork context)
    {
        return false;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    isFailed(AbstractUnitOfWork context)
    {
        return false;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    isCommitted(AbstractUnitOfWork context)
    {
        return false;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    isRolledBack(AbstractUnitOfWork context)
    {
        return false;
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public CompletionStage<Void>
    commit(AbstractUnitOfWork context) 
    {
        return
            CompletableFuture.failedFuture(
                new IllegalStateException(
                    "Cannot call commit() from state: " + getName() + "." ));
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public CompletionStage<Void>
    rollback(AbstractUnitOfWork context)
    {
        return
            CompletableFuture.failedFuture(
                new IllegalStateException(
                    "Cannot call rollback() from state: " + getName() + "." ));
    }

}

// ##########################################################################

File: csharp/Components/Strata.Domain.Core/UnitOfWork/AbstractUnitOfWorkState.cs
//  ##########################################################################
//  # File Name: AbstractUnitOfWorkState.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################

using System;
using System.Collections.Generic;
using System.Linq.Expressions;
using System.Threading.Tasks;
using Strata.Domain.Core.NamedQuery;
using Strata.Foundation.Core.Utility;

namespace Strata.Domain.Core.UnitOfWork
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// $comments$
    /// </summary>
    ///  
    public abstract
    class AbstractUnitOfWorkState:
        IUnitOfWorkState
    {
        public string Name { get; protected set; }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Creates a new <c>AbstractUnitOfWorkState</c> instance.
        /// </summary>
        /// 
        protected
        AbstractUnitOfWorkState(string name)
        {
            Name = name;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public abstract void
        Dispose();

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual Task<E> 
        Insert<K,E>(AbstractUnitOfWork context,E entity) 
            where E: class
        {
            throw 
                new InvalidOperationException(
                    "Cannot call Insert(T) from state: " + Name );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual Task<E> 
        Update<K,E>(AbstractUnitOfWork context,E entity) 
            where E: class
        {
            throw 
                new InvalidOperationException(
                    "Cannot call Update(T) from state: " + Name );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual Task
        Remove<K,E>(AbstractUnitOfWork context,E entity) 
            where E: class
        {
            throw 
                new InvalidOperationException(
                    "Cannot call Remove(T) from state: " + Name );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual Task<Optional<T>> 
        GetUnique<K,T>(AbstractUnitOfWork context,K key) 
            where T: class
        {
            throw 
                new InvalidOperationException(
                    "Cannot call GetUniqueWithKey(K) from state: " + Name );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual Task<Optional<T>> 
        GetUniqueMatching<T>(AbstractUnitOfWork context,
            Expression<Func<T,bool>> predicate) 
            where T: class
        {
            throw 
                new InvalidOperationException(
                    "Cannot call GetUniqueWithPredicate" + 
                    "(Expression<Func<T,bool>>) from state: " + Name );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual Task<IList<T>> GetMatching<T>(AbstractUnitOfWork context,
            Expression<Func<T,bool>> predicate) 
            where T: class
        {
            throw 
                new InvalidOperationException(
                    "Cannot call GetAllWithPredicate" + 
                    "(Expression<Func<T,bool>>) from state: " + Name );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual Task<IList<T>> GetAll<T>(AbstractUnitOfWork context) 
            where T: class
        {
            throw 
                new InvalidOperationException(
                    "Cannot call GetMatching<T>() from state: " + Name );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual Task<INamedQuery<T>> GetNamedQuery<T>(AbstractUnitOfWork context,string queryName) 
            where T: class
        {
            throw 
                new InvalidOperationException(
                    "Cannot call GetNamedQuery<T>(string) from state: " + 
                    Name );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual Task<bool> HasUnique<K,T>(AbstractUnitOfWork context,K key) 
            where T: class
        {
            throw 
                new InvalidOperationException(
                    "Cannot call HasUniqueWithKey(K) from state: " + Name );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual Task<bool> HasUniqueMatching<T>(AbstractUnitOfWork context,
            Expression<Func<T,bool>> predicate) 
            where T: class
        {
            throw 
                new InvalidOperationException(
                    "Cannot call HasUniqueWithPredicate" + 
                    "(Expression<Func<T,bool>>) from state: " + Name );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual Task<bool> HasNamedQuery<T>(AbstractUnitOfWork context,string queryName)
            where T: class
        {
            throw 
                new InvalidOperationException(
                    "Cannot call HasNamedQuery(string) from state: " + Name );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual bool 
        IsActive(AbstractUnitOfWork context)
        {
            return false;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual bool 
        IsFailed(AbstractUnitOfWork context)
        {
            return false;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual bool 
        IsCommitted(AbstractUnitOfWork context)
        {
            return false;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual bool 
        IsRolledBack(AbstractUnitOfWork context)
        {
            return false;
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual Task
            Commit(AbstractUnitOfWork context)
        {
            throw 
                new InvalidOperationException(
                    "Cannot call Commit() from state: " + Name );
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public virtual Task
            Rollback(AbstractUnitOfWork context)
        {
            throw 
                new InvalidOperationException(
                    "Cannot call Rollback() from state: " + Name );
        }
    }
}

//  ##########################################################################
